src/engine/plugins/match.cpp

```diff
@@ -262,6 +264,8 @@ Status MatchPlugin::HandleRequest(const RoutingAlgorithmsInterface &algorithms,
                 "NoMatch", "Requested waypoint parameter could not be matched.", json_result);
         }
     }
+    // we haven't errored yet, only allow leg collapsing if we meet this criteria
+    collapse_legs = sub_matchings.size() == 1 && !parameters.waypoints.empty();
```

Comment:
Maybe it is better to simplify the condition here to `const auto collapse_legs = !parameters.waypoints.empty()` with the following assertion
```
BOOST_ASSERT(parameters.waypoints.empty() || sub_matchings.size() == 1);
const auto collapse_legs = !parameters.waypoints.empty();
```
similar to https://github.com/Project-OSRM/osrm-backend/blob/279c7510cb440e80177ddac1508ec5d13953b06c/include/engine/api/match_api.hpp#L100?

Reply:
Makes sense to me — collapsing should only be contingent upon the original input value of `waypoints`, and the error handling here https://github.com/Project-OSRM/osrm-backend/blob/5756e75c30d5f18d0a03920b44a04e70a5b65e26/src/engine/plugins/match.cpp#L244-L248 should handle the case of a split trace and requested waypoints.